src/dex/stader/stader.ts

```diff
@@ -0,0 +1,240 @@
+import { BI_POWS } from '../../bigint-constants';
+import { AsyncOrSync } from 'ts-essentials';
+import { ETHxEventPool } from './stader-pool';
+import { StaderData, SSPMFunctions } from './types';
+import { StaderConfig } from './config';
+import { WethFunctions } from '../weth/types';
+import ERC20ABI from '../../abi/erc20.json';
+import * as CALLDATA_GAS_COST from '../../calldata-gas-cost';
+import _ from 'lodash';
+import { extractReturnAmountPosition } from '../../executor/utils';
+import { getDexKeysWithNetwork, isETHAddress } from '../../utils';
+
+export class Stader
+  extends SimpleExchange
+  implements IDexTxBuilder<StaderData, any>
+{
+  static dexKeys = ['Stader'];
+  ETHxAddress: string;
+  SSPM_Address: string;
+  SSPMInterface: Interface;
+  StaderOracleAddress: string;
+  StaderOracleInterface: Interface;
+  erc20Interface: Interface;
+  needWrapNative = false;
+  ethxPool: ETHxEventPool;
+  logger: Logger;
+  hasConstantPriceLargeAmounts: boolean = true;
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(_.pick(StaderConfig, ['Stader']));
+
+  constructor(
+    protected network: Network,
+    dexKey: string,
+    dexHelper: IDexHelper,
+    protected config = StaderConfig[dexKey][network],
+  ) {
+    super(dexHelper, 'Stader');
+    this.network = dexHelper.config.data.network;
+    this.ETHxAddress = this.config.ETHx.toLowerCase();
+    this.SSPM_Address = this.config.SSPM.toLowerCase();
+    this.SSPMInterface = new Interface(SSPMAbi as JsonFragment[]);
+    this.StaderOracleAddress = this.config.StaderOracle.toLowerCase();
+    this.StaderOracleInterface = new Interface(
+      StadeOracleAbi as JsonFragment[],
+    );
+    this.erc20Interface = new Interface(ERC20ABI);
+    this.logger = dexHelper.getLogger(this.dexKey);
+    this.ethxPool = new ETHxEventPool(
+      this.dexKey,
+      dexHelper,
+      this.StaderOracleAddress,
+      this.StaderOracleInterface,
+      this.logger,
+    );
+  }
+
+  getAdapterParam(
+    srcToken: Address,
+    destToken: Address,
+    srcAmount: NumberAsString,
+    destAmount: NumberAsString,
+    data: StaderData,
+    side: SwapSide,
+  ): AdapterExchangeParam {
+    return {
+      targetExchange: NULL_ADDRESS,
+      payload: '0x',
+      networkFee: '0',
+    };
+  }
+
+  async initializePricing(blockNumber: number) {
+    await this.ethxPool.initialize(blockNumber);
+  }
+
+  async getPricesVolume(
+    srcToken: Token,
+    destToken: Token,
+    amountsIn: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    limitPools?: string[] | undefined,
+    transferFees?: TransferFeeParams | undefined,
+    isFirstSwap?: boolean | undefined,
+  ): Promise<ExchangePrices<StaderData> | null> {
+    if (side === SwapSide.BUY) return null;
+    if (!this.isEligibleSwap(srcToken, destToken)) return null;
+
+    const pool = this.ethxPool;
+    if (!pool.getState(blockNumber)) return null;
```

Comment:
Does it make sense to implement `getOrGenerateState` on this step? I think we won't lose anything, as this dex is used only for specific token pair

Reply:
I think we can

Reply:
@KanievskyiDanylo fixed